src/adaptation/precond.jl

```diff
@@ -2,50 +2,61 @@
 #### Robust online (co-)variance estimators.
 ####
 
+## Variance estimator
+
 abstract type VarEstimator{T} end
 
 # NOTE: this naive variance estimator is used only in testing
-mutable struct NaiveVar{T} <: VarEstimator{T}
+mutable struct NaiveVar{E<:AbstractVecOrMat{<:AbstractFloat},T<:AbstractVector{E}} <: VarEstimator{T}
     n :: Int
-    S :: Vector{Vector{T}}
+    S :: T
 end
 
 NaiveVar(::Type{T}=Float64) where {T} = NaiveVar(0, Vector{Vector{T}}())
-NaiveVar(::Type{T}, ::Int) where {T} = NaiveVar(T)
-NaiveVar(::Int) = NaiveVar(Float64)
-
-function add_sample!(nc::NaiveVar, s::AbstractVector)
+NaiveVar(::Type{T}, sz::Tuple{Int}) where {T} = NaiveVar(0, Vector{Vector{T}}())
+NaiveVar(::Type{T}, sz::Tuple{Int,Int}) where {T} = NaiveVar(0, Vector{Matrix{T}}())
+# If `sz` are a tuple of two integers, e.g. (10, 2),
+# the adaptor will estimate variance for each column (2 in this case).
+NaiveVar(sz::Tuple{Vararg{Int}}) = NaiveVar(Float64, sz)
+NaiveVar(type::Type{T}, dim::Int) where {T} = NaiveVar(type, (dim,))
+NaiveVar(dim::Int) = NaiveVar((dim,))
+
+function add_sample!(nc::NaiveVar, s::AbstractVecOrMat)
```

Comment:
What do you mean by clarity here?

Reply:
Making the element type explicit help make the code more readable, I think.

Reply:
Yes I agree in general adding types helps readbility. But for this function I feel it's quite self-explained: we add a sample and the same can be either a vector or a matrix. Personally I tend to avoid adding extra type info here because I feel it doesn't help understand reading this function much.